### ISSUE TITLE
Discussion on compiler option naming scheme

### ISSUE THREAD

**Issue by fjeremic** [eclipse/omr#1370]
We have several hundred options in the compiler project [1] and the list is growing week by week. Since consumers of OMR will also depend on these options we should create a discussion on what makes a good option. I had some thoughts to kick things off:

1. Should every option have an "enable" and a "disable" variant?
2. How can we easily identify which options are enabled by default and which are disabled?
3. Is there any way to print documentation for an option?
4. Option documentation is lacking in general. There are many occurences where a "doSomeThing" option is documented with "Do some thing" which is not useful at all.
5. Any thoughts on option naming?
6. What do the various letter prefixes mean in the option comment such as "L", "R", "M", "O", etc?

[1] https://github.com/eclipse/omr/blob/f22653be5e8b92935cb67388f48b08d1de0ab761/compiler/control/OMROptions.cpp#L97-L1301

**Comment by mgaudet** [comment]
1. I think our boolean options should have both. This is an area where a future enhancement to the option system could help by understanding more about our options and automatically creating the variants. 
2. 'default' is a complicated notion, but I'd support anything to help clarification (though, I don't personally have a great answer for this). 
3. `    `
4. `/me nods` -- this is one of those places where it'd be lovely to be able to expand with multiple line strings.
5. `    `
6. `    `

**Comment by JamesKingdon** [comment]
There's a lot to be said for consistency, so that would favour enable/disable prefixes for boolean options. The downside is that as prefixes they are rather long. Not a problem when only using a single option, but I recently found myself having to manage groups of 10 or more options for comparative performance runs, which does start to get quite unwieldy.
I'm not sure that the current status is a good place to pull arguments from, but for the record, for boolean options we currently have examples of enable*, disable*, no* and <no prefix at all>*.
<no prefix> as a synonym for 'enable' is attractive to me on grounds of brevity. A similar argument could be made for using 'no' instead of 'disable', but it looks rather clunky and I think at this stage everyone is used to 'disable' as the natural way of switching a feature off.

I assume that anything we decide here will be applied as guidelines for new options as opposed to being retrospectively applied to the large number of existing options? In which case we will need a pragmatic approach that can coexist with the current rich tapestry of option naming.

**Comment by fjeremic** [comment]
> I assume that anything we decide here will be applied as guidelines for new options as opposed to being retrospectively applied to the large number of existing options? In which case we will need a pragmatic approach that can coexist with the current rich tapestry of option naming.

Yes. We must retain backwards compatibility as some downstream projects may be documenting these options externally. Whatever shape any new option processing facility takes must support aliases so we can keep current option names intact

**Comment by charliegracie** [comment]
Shouldn't option parsing actually be an OMR consumer issue and not an OMR issue?  OMR has lots of configurable options but forcing a consuming runtime to parse specific options does not seem like the right path to me.  Right now we have multiple OMR components being configured in multiple different ways.  I think we should create / design a new configuration story that works for all components and switch to that.  Leave option parsing to the runtime itself as individual runtimes may have very specific ways they want to specify command line options.  OMR *could* provide option parsing helpers if we wanted but I think that these helpers would be used by the runtime itself and still feed into whatever configuration story we came up with.

My 2 cents

**Comment by mstoodle** [comment]
Kicking the can down the road to the consuming runtime has its attractions, but also disadvantages for anyone who works on more than one runtime. One of the advantages of a single technology base, like Eclipse OMR, underpinning multiple runtimes is supposed to be helping to unify how people, tools, frameworks, and platforms can configure and interact with those runtimes in a consistent way. If the runtime decides the interface, then there's no common footing.

I remember earlier compiler technology sharing with this code base where I'm not sure there was a single person who knew how to generate a compiler log across all the language compilers because they all had their own way to specify it (granted, varying platforms was part of the issue there). I would like to believe we can create a better story than that.

I think we should be defining an options API at the OMR project level that can both be used directly and adapted (or fed) by existing runtimes that have their own options processing logic already. Possibly the runtime could either allow (by default) or prevent (by explicit action) the common API from being exposed to give the option to "close the back door", so to speak, that the OMR API would open up. Ideally, all the OMR components would fit underneath that same API.

I like the suggestions that have been made so far.

How do we appropriately future proof options so that we have the flexibility to evolve in ways that benefit our community without flagrantly breaking our consumers on a regular basis? Perhaps each option could be associated with the semantic version level that has to bump to let the option change? i.e. enabling/disabling some small transformation used only to diagnose a bug in a particular opt could be allowed to change with less fanfare than, for example, more fundamental operations that configure the size of the JIT code cache or to configure the nursery size in the heap. We may want to only change some options if there's a migration path provided at the same time so downstream projects can manage the transition.

Just throwing out ideas...we could do elaborate things here that are probably not (yet?) warranted or we could do nothing new and see how bad that gets. But I think there's some useful intermediate ground that would establish some dependable ground rules.

**Comment by fjeremic** [comment]
Linking this to #2835 for reference. I think we can go ahead and close this issue though?
